### ISSUE TITLE
GPL-559	Automatically batch lighthouse sentinel samples

### ISSUE THREAD

**Issue by rl15** [sanger/crawler#56]
**User story**
GPL-559 | As genomics POs (Emma, Rich C & James) we would like Lighthouse sentinel samples to be automatically batched for cherry picking

**Who are the primary contacts for this story**
Emma G
Rich C
James M
Danni W

**Acceptance criteria*

Acceptance criteria TBD after bulk cherry picking UAT complete (Operations need to understand new process first)

To be considered successful the solution must allow:
- [ ] add a list of acceptance criteria here
- [ ] ...

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

**Additional context**
See sequence flow diagram: https://ssg-confluence.internal.sanger.ac.uk/x/jqsKBg

**Comment by rl15** [comment]
Mockup (mailed to SSR Tuesday, 14 July 2020 at 11:20): 
[20200714_GPL-559_Batch_Sentinal_CP.pdf](https://github.com/sanger/crawler/files/4918203/20200714_GPL-559_Batch_Sentinal_CP.pdf)

**Comment by JamesGlover** [comment]
This is a bit of a diversion from the current process for generating batches. Its one of the reasons I had hoped to get cherrypicking re-written at some point.

Hidden requirements:
- Sequencescape batches are collections of requests. So there will be a need to build requests.
- Requests are usually built via a submission/order. Given the vast majority of requests in Sequencescape (and all cherrypick requests) have a submission, this may be risky to bypass.
- Requests need some additional information: Study, Project, Request Type

Infrastructure in place:
- Submission templates provide API endpoint (currently just V1) to automatically build submissions
- This could automatically build the requests as needed, although its a little harder to add and remove requests after built.
- Submissions templates can have different builder classes (eg. To automatically build a batch)
- Its also possible fro requests to register a callback with the submission, which gets executed after the submission is built (example below) however this would require a different request class for lighthouse picks compared to normal cherrypick requests.
```
class Request::AutoMultiplexing < Request::Multiplexing
  after_create :register_transfer_callback

  # Triggers immediate transfer into the tubes if the source asset already
  # exists. This allows multiplexing requests to be made on plates at the
  # end of library prep, after the plate is qc_complete.
  # If no asset is present then we haven't got to that stage yet and transfer
  # will be triggered as part of the standard workflow.
  def register_transfer_callback
    # We go via order as we need to get a particular instance of submission
    return if asset.blank?

    order.submission.register_callback(:once) do
      Transfer::FromPlateToTubeByMultiplex.create!(
        source: asset.plate,
        user: order.user
      )
    end
  end
end
```



Other things to consider:
- Submission building is mostly asynchronous, which would be undesirable if you want immediate

**Comment by JamesGlover** [comment]
I'd suggest adding a 'Batched cherrypick' submission template, and then trigger that via the API. This would have the advantage that it would also expose the functionality within Sequencescape in general. That way we'd need:

- SubmissionTemplate should generate a 'BatchedLinearSubmission'
- BatchedLinear submission use standard submission behaviour, but batch up all requests of the first request type automatically.
- Expose the submission template endpoints via the V2 API (probably)
- Add a synchronous: true flag to submission to build them synchronously.
- Submissions `have_many :batches, via: requests` and then expose thos batches via the API

**Comment by JamesGlover** [comment]
OOh! Better idea!
New endpoint: `pick-list` which just takes an array of receptacles, and study information, and does all this under the hood. That way if we get around to substantially changing the way cherrypicking works we can do so in a nicely isolate manner.

**Comment by emrojo** [comment]
I think this feature (create a batch automatically) is mainly a change in SS so is going to be implemented in https://github.com/sanger/sequencescape/issues/2837 , I dont see what should be changed in Lighthouse/Crawler. 
I'll add the 'On hold' flag for this reason.

**Comment by JamesGlover** [comment]
> I think this feature (create a batch automatically) is mainly a change in SS so is going to be implemented in [sanger/sequencescape#2837](https://github.com/sanger/sequencescape/issues/2837) , I dont see what should be changed in Lighthouse/Crawler.
> I'll add the 'Question' flag for this reason.

There is the front end component in the app @stevieing is delivering as part of GPL-558 (sanger/lighthouse-ui#2) which will mostly be the API call to Sequencescape, and displaying the returned link.
